packages/cli/src/config/settings.ts

```diff
@@ -640,11 +658,30 @@ export function loadSettings(
   // the settings to avoid a cycle
   loadEnvironment(tempMergedSettings);
 
-  // Now that the environment is loaded, resolve variables in the settings.
+  // Now that the environment is loaded, track and resolve variables in the settings.
+  // Track environment variable mappings before resolution
+  const systemOriginal = structuredClone(systemSettings);
```

Comment:
the amount of logic duplicated for each of the settings files is now too high. Please add a helper that does all the steps duplicated for each of the env